Mathlib/Tactic/Widget/StringDiagram.lean

```diff
@@ -0,0 +1,543 @@
+  | .id f, .of g => do
+    return .id (f.comp (.of g))
+  | .cons (.of η) θ, .of f => do
+    let η' := WhiskerRightExpr.whisker η f
+    let θ' ← evalWhiskerRightExpr θ (.of f)
+    return .cons (.of η') θ'
+  | .cons (.whisker f η) θ, .of h => do
+    let g ← η.src
+    let g' ← η.tar
+    let η' ← evalWhiskerLeftExpr (.of f) (← evalWhiskerRightExpr (.cons η (.id g')) (.of h))
+    let θ' ← evalWhiskerRightExpr θ (.of h)
+    return evalComp (← NormalExpr.associator (.of f) g (.of h)) (evalComp η' (evalComp (← NormalExpr.associatorInv (.of f) g' (.of h)) θ'))
+  | η, .comp g h => do
+    let η' ← evalWhiskerRightExpr (← evalWhiskerRightExpr η g) h
+    let f ← η.src
+    let f' ← η.tar
+    return evalComp (← NormalExpr.associatorInv f g h) (evalComp η' (← NormalExpr.associator f' g h))
+  | η, .id _ => do
+    let f ← η.src
+    let g ← η.tar
+    return evalComp (← NormalExpr.rightUnitor f) (evalComp η (← NormalExpr.rightUnitorInv g))
+
+def WhiskerRightExpr.e : WhiskerRightExpr → MetaM Expr
+  | WhiskerRightExpr.of η => η.e
+  | WhiskerRightExpr.whisker η f => do
+    mkAppM ``MonoidalCategoryStruct.whiskerRight #[← η.e, f.e]
+
+def WhiskerLeftExpr.e : WhiskerLeftExpr → MetaM Expr
+  | WhiskerLeftExpr.of η => η.e
+  | WhiskerLeftExpr.whisker f η => do
+    mkAppM ``MonoidalCategoryStruct.whiskerLeft #[f.e, ← η.e]
+
+def NormalExpr.e : NormalExpr → MetaM Expr
+  | NormalExpr.id f => do mkAppM ``CategoryStruct.id #[← f.e]
+  | NormalExpr.cons η (NormalExpr.id _) => η.e
+  | NormalExpr.cons η θ => do mkAppM ``CategoryStruct.comp #[← η.e, ← θ.e]
+
+def NormalExpr.toList : NormalExpr → List WhiskerLeftExpr
+  | NormalExpr.id _ => []
+  | NormalExpr.cons η θ => η :: NormalExpr.toList θ
+
+def WhiskerRightExpr.core : WhiskerRightExpr → Core
+  | WhiskerRightExpr.of η => η
+  | WhiskerRightExpr.whisker η _ => η.core
+
+def WhiskerLeftExpr.core : WhiskerLeftExpr → Core
+  | WhiskerLeftExpr.of η => η.core
+  | WhiskerLeftExpr.whisker _ η => η.core
+
+def WhiskerRightExpr.structural? (η : WhiskerRightExpr) : Option WhiskerRightExpr :=
+  match η.core with
+  | .of _ => none
+  | .ofStructural _ => some η
+
+def WhiskerLeftExpr.structural? (η : WhiskerLeftExpr) : Option WhiskerLeftExpr :=
+  match η.core with
+  | .of _ => none
+  | .ofStructural _ => some η
+
+def NormalExpr.structural? : NormalExpr → Option NormalExpr
+  | NormalExpr.id f => some (.id f)
+  | NormalExpr.cons η θ =>
+    match η.structural?, θ.structural? with
+    | some _, some _ => some (.cons η θ)
+    | _, _ => none
+
+/-- Interpret an `Expr` term as a `Core` term. -/
+def toCore (e : Expr) : Core :=
+  match structuralAtom? e with
+  | some η => Core.ofStructural η
+  | none => Core.of ⟨e⟩
+
+def NormalExpr.of (η : Expr) : MetaM NormalExpr := do
+  return .cons (.of (.of (toCore η))) (.id <| ← StringDiagram.tar η)
+
+/-- Evaluate the expression of a 2-morphism into a normalized form. -/
+partial def eval (e : Expr) : MetaM NormalExpr := do
```

Comment:
Can you add some tests for this in the test file? I think the tests will be easiest with Qq, as something like:
```
variable {u v : Level} {C : Q(Type u)} {iC : Q(Category.{v} $C)} {iM : Q(MonoidalCategory $C)}
variable {X Y : Q($C)} (η : Q(𝟙_ _ ⟶ $X ⊗ $Y)) (ε : Q($Y ⊗ $X ⟶ 𝟙_ _))

#guard_msgs
#eval eval q(η ▷ X ≫ (α_ _ _ _).hom ≫ X ◁ ε)
```